force-app/classes/src/Assertions.cls

```diff
@@ -35,72 +35,156 @@ public class Assertions {
         this.spy.hasBeenCalled(),
-        'Method ' +
-        this.spy.methodName +
-        ' was not called'
+        buildErrorMessage(this.spy, ErrorMessageType.NEVER_CALLED)
       );
       this.asserter.assertEquals(
         true,
         this.spy.hasBeenLastCalledWith(params),
-        'Method ' +
-        this.spy.methodName +
-        ' was not last called with ' +
-        params
+        buildErrorMessage(
+          this.spy,
+          ErrorMessageType.NOT_LAST_CALLED_WITH,
+          params
+        )
       );
     }
   }
 
   public static MethodSpyAssert assertThat(MethodSpy spy) {
     return new MethodSpyAssert(spy);
   }
+
+  private enum ErrorMessageType {
+    CALLED,
+    NEVER_CALLED,
+    NOT_CALLED_WITH,
+    NOT_LAST_CALLED_WITH,
+    NOT_CALLED_TIMES
+  }
+
+  private class ErrorMessageRow {
+    protected String message;
+    protected Boolean tabbed;
+
+    ErrorMessageRow(String message) {
+      this(message, false);
+    }
+    ErrorMessageRow(String message, Boolean tabbed) {
+      this.message = message;
+      this.tabbed = tabbed;
+    }
+  }
+
+  private class ErrorMessage {
+    private String headline;
+    private List<ErrorMessageRow> callTraces;
+    ErrorMessage(MethodSpy spy, ErrorMessageType error) {
+      this(spy, error, null);
+    }
+    ErrorMessage(MethodSpy spy, ErrorMessageType error, Object argument) {
+      this.headline = this.buildHeadline(spy, error, argument);
+      this.callTraces = this.buildCallTraces(spy);
+    }
+
+    private String buildHeadline(
+      MethodSpy spy,
+      ErrorMessageType error,
+      Object argument
+    ) {
+      final Map<ErrorMessageType, String> messages = new Map<ErrorMessageType, String>{
+        ErrorMessageType.CALLED => 'Method {0} was called',
+        ErrorMessageType.NEVER_CALLED => 'Method {0} was not called',
+        ErrorMessageType.NOT_CALLED_WITH => 'Method {0} was not called with {1}',
+        ErrorMessageType.NOT_LAST_CALLED_WITH => 'Method {0} was not last called with {1}',
+        ErrorMessageType.NOT_CALLED_TIMES => 'Method {0} was not called {1} times'
+      };
+      return String.format(
+        messages.get(error),
+        new List<String>{ spy.methodName, argument + '' }
+      );
+    }
+
+    private List<ErrorMessageRow> buildCallTraces(MethodSpy spy) {
+      final List<Params> callLogParams = spy.getCallLogParams();
+      if (callLogParams.isEmpty()) {
+        return new List<ErrorMessageRow>();
+      }
+
+      final List<ErrorMessageRow> rows = new List<ErrorMessageRow>();
+      rows.add(new ErrorMessageRow('method call history (latest first):'));
+      for (Params params : callLogParams) {
+        rows.add(
+          new ErrorMessageRow(spy.methodName + '(' + params + ')', true)
+        );
+      }
+      return rows;
+    }
+
+    override public String toString() {
+      final List<String> lines = new List<String>();
+      lines.add(this.headline);
+      if (!callTraces.isEmpty()) {
```

Comment:
the goal here is to avoid adding `method call history:` when we don't have anything to display. Feel free to explain a bit more about your comment if I misunderstood it.

Reply:
Ah I see, exact
Thanks, you understood it correctly, I read the code too fast and miss the first "header" row